File: src/srv_test.py
# -*- coding: UTF-8 -*-

import unittest
import logging
import json

import webtest

import srv


class SrvTest(unittest.TestCase):

    def test_root(self):

        tests = [
            {"in": '{"name":"Monkey","age":10}', "out": "Monkey's age is 10\n", "code": 200},
            {"in": '{"name":"Monkey","age":10.1}', "out": "Monkey's age is 10\n", "code": 200},
            {"in": '{"name":"Monkey","age":"10"}', "out": "400 Bad Request: 'age' must be a number\n", "code": 400},
            {"in": '{"name":"Monkey"}', "out": "400 Bad Request: 'age' is required\n", "code": 400},
            {"in": '', "out": "400 Bad Request: No JSON object could be decoded\n", "code": 400},
        ]

        app = webtest.TestApp(srv.app)

        for test in tests:
            response = app.post('/', test["in"], status=test["code"])
            self.assertEqual(response.body, test["out"])
            self.assertEqual(response.status_int, test["code"])


if __name__ == "__main__":

    logging.basicConfig(level=logging.CRITICAL)
    unittest.main()

File: src/srv_test.go
package main

import (
	"io/ioutil"
	"net/http"
	"net/http/httptest"
	"testing"
	"strings"
)

func TestRoot(t *testing.T) {

	ts := httptest.NewServer(nil)
	defer ts.Close()

	tests := []struct{
		in string
		out string
		status int
	}{
		{`{"name":"Monkey","age":10}`, "Monkey's age is 10\n", 200},
		{`{"name":"Monkey","age":10.1}`, "Monkey's age is 10\n", 200},
		{`{"name":"Monkey","age":"10"}`, "400 Bad Request: 'age' must be a number\n", 400},
		{`{"name":"Monkey"}`, "400 Bad Request: 'age' is required\n", 400},
		{"", "400 Bad Request: unexpected end of JSON input\n", 400},
	}

	for _, test := range tests   {

		res, err := http.Post(ts.URL, "application/json", strings.NewReader(test.in))
		if err != nil {
			t.Fatal(err)
		}

		body, err := ioutil.ReadAll(res.Body)
		res.Body.Close()
		if err != nil {
			t.Fatal(err)
		}

		if string(body) != test.out {
			t.Fatalf("expected '%s', got '%s'", test.out, string(body))
		}

		if res.StatusCode != test.status {
			t.Fatalf("expected code %d, got %d", test.status, res.StatusCode)
		}

	}

}
